### PR TITLE
Upgrade to Transformers v4.45

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

As per title.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
@regisss should we loosen the requirements for https://github.com/huggingface/optimum-habana/blob/8043d2cef69edc9eae6c7282bbb7fa41f268e5b6/examples/language-modeling/requirements.txt#L7C1-L7C15 to `peft>=0.11.1,<=0.13.0`

**Comment by regisss** [comment]
> @regisss should we loosen the requirements for https://github.com/huggingface/optimum-habana/blob/8043d2cef69edc9eae6c7282bbb7fa41f268e5b6/examples/language-modeling/requirements.txt#L7C1-L7C15 to `peft>=0.11.1,<=0.13.0`

We can do it. Is the current constraint too strict for some examples?

**Comment by harborn** [comment]
when will this PR merge?
currently, I need this PR to use transformers==4.45.0

**Comment by regisss** [comment]
> when will this PR merge? currently, I need this PR to use transformers==4.45.0

When all our internal tests are validated. Probably next week but guarantee it.
In the meantime, feel free to install this branch directly.

**Comment by regisss** [comment]
> > > when will this PR merge? currently, I need this PR to use transformers==4.45.0
> > 
> > 
> > When all our internal tests are validated. Probably next week but guarantee it. In the meantime, feel free to install this branch directly.
> 
> I have pulled this patch and run my workflow based on this patch to inference llama-3.2-11b and llama-3.2-90b, it can work, but with lower performance.

Lower performance means lower throughput than Llama 3.1?

**Comment by yafshar** [comment]
> > @regisss should we loosen the requirements for https://github.com/huggingface/optimum-habana/blob/8043d2cef69edc9eae6c7282bbb7fa41f268e5b6/examples/language-modeling/requirements.txt#L7C1-L7C15 to `peft>=0.11.1,<=0.13.0`
> 
> We can do it. Is the current constraint too strict for some examples?

Some of the dependencies like `neural-compressor` will cause a different version installation and when you are in the language modeling example with the hard constraint for peft, which I do not see the reason, it will cause re-installation and some pip dependencies  error + warnings  . @sywangyi is there a constraint that you fixed the version to 0.12.0 or can it be loose as above?

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -1302,11 +1304,18 @@ def forward(
...
+                    "Starting from v4.46, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
+                )
+            # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+            # TODO: remove the float() operation in v4.46
+            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
 
         loss = None
         if labels is not None:
+            # Upcast to float if we need to compute the loss to avoid potential precision issues
+            logits = logits.float()
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:1318]: why this is need to be logit float?

File: optimum/habana/transformers/models/bloom/modeling_bloom.py [hunk 2]
```diff
@@ -498,14 +504,18 @@ def prepare_inputs_for_generation(
...
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and past_key_values is None:
-            model_inputs = {"inputs_embeds": inputs_embeds}
+            model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
         else:
-            model_inputs = {"input_ids": input_ids}
+            # This `clone` call is needed to avoid recapturing cuda graphs with `torch.compile`'s  `mode="reduce-overhead`, as otherwise the
+            # input `position_ids` would have various stride during the decoding. Here, simply using `.contiguous()` is not sufficient as in
+            # the batch size = 1 case, `position_ids` is already contiguous but with varying stride which retriggers a capture.
+            model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format), "inputs_embeds": None}
```
Comment (libinta) [inline optimum/habana/transformers/models/bloom/modeling_bloom.py:512]: add if lazy_mode check

File: optimum/habana/transformers/models/codegen/modeling_codegen.py [hunk 3]
```diff
@@ -343,17 +359,21 @@ def prepare_inputs_for_generation(
                 else:
                     position_ids = position_ids[:, -1]
 
+                # This `clone` call is needed to avoid recapturing cuda graphs with `torch.compile`'s  `mode="reduce-overhead`, as otherwise the input `position_ids` would have various stride during the decoding. Here, simply using `.contiguous()` is not sufficient as in the batch size = 1 case, `position_ids` is already contiguous but with varying stride which retriggers a capture.
+                position_ids = position_ids.clone(memory_format=torch.contiguous_format)
```
Comment (libinta) [inline optimum/habana/transformers/models/codegen/modeling_codegen.py:363]: add lazy_mode check

File: optimum/habana/transformers/models/codegen/modeling_codegen.py [hunk 4]
```diff
@@ -343,17 +359,21 @@ def prepare_inputs_for_generation(
...
+
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and past_key_values is None:
             model_inputs = {"inputs_embeds": inputs_embeds}
         else:
-            model_inputs = {"input_ids": input_ids.contiguous()}
+            model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format)}
```
Comment (libinta) [inline optimum/habana/transformers/models/codegen/modeling_codegen.py:369]: same here

File: optimum/habana/transformers/models/phi/modeling_phi.py [hunk 5]
```diff
@@ -575,11 +576,18 @@ def forward(
...
+                "Starting from v4.46, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
+            )
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        # TODO: remove the float() operation in v4.46
+        logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
 
         loss = None
         if labels is not None:
+            # Upcast to float if we need to compute the loss to avoid potential precision issues
+            logits = logits.float()
```
Comment (libinta) [inline optimum/habana/transformers/models/phi/modeling_phi.py:590]: dup of line585?

**Comment by sywangyi** [comment]
> > > @regisss should we loosen the requirements for https://github.com/huggingface/optimum-habana/blob/8043d2cef69edc9eae6c7282bbb7fa41f268e5b6/examples/language-modeling/requirements.txt#L7C1-L7C15 to `peft>=0.11.1,<=0.13.0`
> > 
> > 
> > We can do it. Is the current constraint too strict for some examples?
> 
> Some of the dependencies like `neural-compressor` will cause a different version installation and when you are in the language modeling example with the hard constraint for peft, which I do not see the reason, it will cause re-installation and some pip dependencies error + warnings . @sywangyi is there a constraint that you fixed the version to 0.12.0 or can it be loose as above?

I just upgrade peft to the latest tag when I enable boft, ln_tuning and vera. this the peft test could pass , I am ok to loss as above. test is in https://github.com/huggingface/optimum-habana/blob/main/tests/test_peft_inference.py, and https://github.com/huggingface/optimum-habana/blob/main/tests/test_examples.py

**Comment by harborn** [comment]
I see that tranformers==4.45.1 released, so any changes to upgrade again if use transformers==4.45.1 ?

File: optimum/habana/transformers/generation/stopping_criteria.py [hunk 6]
```diff
@@ -52,18 +52,6 @@ def gaudi_MaxLengthCriteria_call(
...
-def gaudi_MaxNewTokensCriteria_call(
-    self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs
-) -> Union[torch.BoolTensor, bool]:
-    token_idx = kwargs.get("token_idx", None)
-    if token_idx is not None:
-        assert not kwargs["needs_tensor_output"]
-        return token_idx >= self.max_length
-    else:
-        is_done = input_ids.shape[-1] >= self.max_length
-        return create_return_const_tensor(input_ids, is_done)
```
Comment (ssarkar2) [inline optimum/habana/transformers/generation/stopping_criteria.py:64]: @libinta MaxNewTokensCriteria no longer exists in transformers.

removed in this PR: https://github.com/huggingface/transformers/pull/32659/files#diff-6e63ae0764aa864afd5bae6d512677b99b5240cb98cb210190482bdbb6a85906


It was removed as it had plans for being deprecated:

"The class `MaxNewTokensCriteria` is deprecated and will be removed in v4.43. "
            f"Please use `MaxLengthCriteria(max_length={start_length + max_new_tokens})` "

File: optimum/habana/transformers/models/phi/modeling_phi.py (hunk 5 above)
Comment (regisss) [reply optimum/habana/transformers/models/phi/modeling_phi.py:590]: It seems the `.float()` of line 585 will be removed in Transformers v4.46. Do you see any perf degradation?

File: optimum/habana/transformers/models/codegen/modeling_codegen.py (hunk 3 above)
Comment (libinta) [reply optimum/habana/transformers/models/codegen/modeling_codegen.py:363]: i think clone causes perf issue, and we dont need this if it's not torch,compile,right?
